content-services/content-sources-backend
# Fixes 3255: support pulp content guards

## Summary

When snapshotting a custom repository, we check to see if there are three content guards created in the org:
1.  content guard that checks for org_id on the identity header
2.  content guard that checks for a specific custom_dn on the identity header (used by turnpike)
3. content guard that Composites these together

If any of these 3 are not created, or not configured correctly, they will be updated or created.  The created pulp distribution is then associated with the content guard.

## Testing steps

1. Set config values for `guard_subject_dn` and `custom_repo_content_guards`
```
  pulp:
    download_policy: on_demand
    server: http://localhost:8080
    username: admin
    password: password
    storage_type: local #object or local
    guard_subject_dn: "some_dn"
    custom_repo_content_guards: true
```

you can change 'some_dn' to whatever you want, but remember what you set it to!

5. Import a small redhat repo:
``` 
OPTIONS_REPOSITORY_IMPORT_FILTER=small go run cmd/external-repos/main.go import
```

6.  Run the server:  ```make run```

3.  Trigger sync:  ``` go run cmd/external-repos/main.go nightly-jobs```
7.  create a custom repo with snapshotting, in the UI and let it snapshot
8.  Grab the org id from the db (ignore -1):  

`make db-cli-connect`  
```
select org_id from where repository_configurations;
```

9.  Fetch the Red Hat repo without anything specified.  Navigate to the red hat repos, you should have one repo, list snapshots and grab the repo config.  Grab the URL for the repository.  For local development, change the hostname and port to localhost:8080. for example:

http://localhost:8080/pulp/content/abcde/fghjkl/

You should be able to curl that url and get valid html (200 Response), as well as fetch its ./repodata/repomd.xml
```
curl http://localhost:8080/pulp/content/abcde/fghjkl/repodata/repomd.xml
```

10. Get unauthorized when fetching the custom repo.  Similar to the redhat repo, grab the snapshot URL  and change the host/port and try to curl it, you should get an unauthorized response
11.  Now fetch using a cert with a proper org id with your org_id:
```
curl  http://localhost:8080/pulp/content/abcde/fghjkl/repodata/repomd.xml   -H "`./scripts/header.sh $ORG_ID 123 jdoe`"
```
12.  Now fetch using a 'turnpike' header:

```
curl  http://localhost:8080/pulp/content/abcde/fghjkl/repodata/repomd.xml   -H "`./scripts/turnpike_header.sh $some_dn`"
```
replacing some_dn with whatever you put in your config at step 1

## Checklist
